Src/IronPython/Runtime/Operations/FloatOps.cs

```diff
@@ -1017,6 +1017,29 @@ public static partial class DoubleOps {
                     throw PythonOps.ValueError("__setformat__() argument 1 must be 'double' or 'float'");
             }
         }
+
+        public static double __round__(double self) {
+            // as of Python 3 rounding is to the nearest even number, not away from zero
+            return MathUtils.Round(self, 0, MidpointRounding.ToEven);
+        }
+
+        public static double __round__(double self, int ndigits) {
+            return PythonOps.CheckMath(self, MathUtils.Round(self, ndigits, MidpointRounding.ToEven));
+        }
+
+        public static double __round__(double self, BigInteger ndigits) {
+            int n;
+            if (ndigits.AsInt32(out n)) {
+                return __round__(self, n);
+            }
+
+            return ndigits > 0 ? self : 0.0;
+        }
+
+        public static double __round__(double self, object ndigits) {
```

Comment:
Not that I necessarily care about supporting this (very) edge case, but the following is valid:
```Python
class foo(object):
  def __index__(self):
    return 0

assert float.__round__(1.2, foo()) == 1.0
```
Same comment in `IntOps` and `LongOps`.

Reply:
No this doesn't work. But it isn't clear to me what the best way to deal with this is.

I could explicitly look for an __index__ member and call it, but this seems like it could be a specific case of a more general phenomenon.

Is there a way within the C# code to attempt to cast a python object to an int / long using the same logic as the runtime would use?

Reply:
I saw your other post about error messages, so I tried commenting out the `object ndigits` overload, expecting that this would fix the issue, however it doesn't.

I would have expected this case to be handled by the binding mechanism. i.e. `PythonOps.CallWithContextAndThis` should understand that an object with a special `__index__` member can satisfy the int parameter and perform the conversion.

What do you think?

Reply:
We can't just leave it to the binding mechanism since I don't think CPython always does the conversion, for example `int(foo())` is not valid.

Seems like having some sort of attribute (e.g. `[AllowIndex]int bla`) that the binding could pick up may be interesting, but I don't know how that might affect performance. Anyway, it's not necessarily a trivial change.

There are probably a bunch of different ways that it's done in the codebase. An option would be something like `ndigits = PythonOps.Index(ndigits)`. I believe it will return an int/bigint so you'd still need to check the type and dispatch properly...

Reply:
OK, this looks doable. Actually it explains the error message I was trying to replicate. In fact this seems to be exactly what Python 3 is doing.